tools/ci/rules.go

```diff
@@ -21,3 +21,47 @@ func createHttpClient(m dsl.Matcher) {
 		`new(http.Transport)`,
 	).Report(`http.Client: use fleethttp.NewClient instead`)
 }
+
+func txCheck(m dsl.Matcher) {
+	m.Import("github.com/fleetdm/fleet/v4/server/datastore/mysql")
+	m.Import("github.com/jmoiron/sqlx")
+
+	m.Match(`$ds.withTx($_, $fn)`, `$ds.withRetryTxx($_, $fn)`).
+		Where(m["ds"].Type.Is(`*mysql.Datastore`) && (m["fn"].Contains(`$ds.writer`) || m["fn"].Contains(`$ds.reader`))).
+		Report("improper use of ds.reader or ds.writer in a transaction")
+
+	// any Datastore method that receives a Tx (sqlx.ExtContext/sqlx.ExecContext)
+	// as the first argument must use it.
+	m.Match(`func ($ds *Datastore) $name($p1, $*_) $*_ { $*fn }`, `func ($ds Datastore) $name($p1, $*_) $*_ { $*fn }`).
+		Where(
+			(m["ds"].Type.Is(`*mysql.Datastore`) || m["ds"].Type.Is(`mysql.Datastore`)) &&
+				(m["p1"].Type.Is(`sqlx.ExtContext`) || m["p1"].Type.Is(`sqlx.ExecContext`)) &&
+				(m["fn"].Contains(`$ds.writer`) || m["fn"].Contains(`$ds.reader`))).
+		Report("improper use of ds.reader or ds.writer in Datastore.$name")
+
+	// any Datastore method that receives a Tx (sqlx.ExtContext/sqlx.ExecerContext)
+	// as the second argument must use it (e.g. ds.method(ctx, tx, ...)).
+	m.Match(`func ($ds *Datastore) $name($_, $p2, $*_) $*_ { $*fn }`, `func ($ds Datastore) $name($_, $p2, $*_) $*_ { $*fn }`).
+		Where(
+			(m["ds"].Type.Is(`*mysql.Datastore`) || m["ds"].Type.Is(`mysql.Datastore`)) &&
+				(m["p2"].Type.Is(`sqlx.ExtContext`) || m["p2"].Type.Is(`sqlx.ExecerContext`)) &&
+				(m["fn"].Contains(`$ds.writer`) || m["fn"].Contains(`$ds.reader`))).
+		Report("improper use of ds.reader or ds.writer in Datastore.$name")
+
+	// any func literal that receives a Tx (sqlx.ExtContext/sqlx.ExecerContext)
+	// as the first argument must use it.
+	m.Match(`func($p1, $*_) $*_ { $*fn }`).
+		Where(
+			(m["p1"].Type.Is(`sqlx.ExtContext`) || m["p1"].Type.Is(`sqlx.ExecerContext`)) &&
+				(m["fn"].Contains(`$ds.writer`) || m["fn"].Contains(`$ds.reader`))).
+		Report("improper use of ds.reader or ds.writer in function literal")
+
+	// TODO: https://github.com/fleetdm/fleet/pull/8621#pullrequestreview-1172676063
+	// This misses the case where a call to a Datastore method is done in one of
+	// those functions and that method uses the reader/writer (and does not
+	// receive a Tx as argument).
+	//
+	// I don't think the ruleguard pattern-matching syntax supports such a case
+	// (recursively check if the function and any of its callees use some field),
+	// it would probably require using the lower-level go/analysis package.
```

Comment:
Couldn't figure out how to do it with ruleguard (I don't think it's possible), left a note as potential improvement.

Reply:
Could we check that all the function calls that are `ds.` take a tx param?

Reply:
> Could we check that all the function calls that are ds. take a tx param?

Possibly... But I'm not sure how to nest a matcher in a matcher (i.e. this would be a check done _inside_ the `fn` var from the existing matchers). I'll do another check but I don't want to spend too much time on this.

Reply:
Certainly, it was just a thought.

Reply:
Yeah the problem is that in `containsIllegal` above, I cannot do other matching, the only available operations are `Contains(pattern string) bool` and `Filter` that is similar but with a func. I don't see how we can run more matchers with conditions (with `Where`) based on a var captured in a parent matcher (`fn` in our case).

Reply:
Well, if this limits us to only have to check this particular case only, that's a win regardless.